@here/olp-sdk-core/lib/utils/MultiPartUploadWrapper.ts

```diff
@@ -177,9 +221,211 @@ export class MultiPartUploadWrapper {
     ): Promise<number> {
+        const defaultParallelRequests = 6;
+        const parallelRequests =
+            opts.parallelRequests || defaultParallelRequests;
+
+        const parts: { id: string; number: number }[] = [];
+
+        const blobData = await this.prepareBlobData(data);
+
+        const dataSize = blobData.size();
+
+        const request = await this.getUploadRequest({
+            blobVersion: opts.blobVersion,
+            catalogHrn: opts.catalogHrn,
+            abotrSignal
+        });
+
+        const startMultipartResponse = await request.startMultipartUpload({
+            contentType: opts.contentType,
+            handle: opts.handle,
+            layerId: opts.layerId,
+            billingTag: opts.billingTag,
+            contentEncoding: opts.contentEncoding
+        });
+
+        if (!this.isStartMultipartResponseValid(startMultipartResponse)) {
+            throw new Error(
+                "Failed to start multipart upload process, bad response."
+            );
+        }
+
+        opts.onStart &&
+            opts.onStart({
+                dataSize,
+                multipartToken: startMultipartResponse.multipartToken,
+                multipartUrl: startMultipartResponse.uploadPartUrl,
+                multipartStatusUrl: startMultipartResponse.statusUrl
+            });
+
+        const chunkSize = this.validateChunkSize(opts.chunkSizeMB);
+        const totalChunks = Math.ceil(dataSize / chunkSize);
+
+        let unreadBytes = dataSize;
+        let offset = 0;
+        let chunkNumber = 0;
+        let uploadedChunks = 0;
+
+        while (unreadBytes > 0) {
+            const promises = [];
+
+            while (promises.length < parallelRequests) {
+                const buffer = await blobData.readBytes(offset, chunkSize);
+
+                if (buffer.byteLength === 0) {
+                    break;
+                }
+
+                chunkNumber++;
+                offset += buffer.byteLength;
+                unreadBytes -= buffer.byteLength;
+
+                promises.push(
+                    request
+                        .uploadPart({
+                            data: buffer,
+                            layerId: opts.layerId,
+                            multipartToken:
+                                startMultipartResponse.multipartToken,
+                            billingTag: opts.billingTag,
+                            contentLength: buffer.byteLength,
+                            contentType: opts.contentType,
+                            partNumber: chunkNumber,
+                            url: startMultipartResponse.uploadPartUrl
+                        })
+                        .then(uploadPartResponse => {
+                            parts.push({
+                                id: uploadPartResponse.id,
+                                number: chunkNumber
+                            });
+
+                            uploadedChunks++;
```

Comment:
Is this thread-safe? Incrementing a number from multiple threads is risky and will probably result eventually in false incremental value.

Reply:
It's one thread. Chunks uploads in an async way and after each uploaded chunk, we increment the counter. We don't care about the order of uploaded chunks. We just need to know the total uploaded chunks. The user will get the actual count of uploaded chunks in each event.

Reply:
I am still confused. How can it be in parallel and then increment the counter in a serial way? So the uploads happen in another thread/context but the then() continuation is called one by one for these 6 uploads?

Reply:
We create the first array of 6 promises and each promise will increment in own then() the counter after will be bone. Those 6 promises not running immediately and starting on line 353 `await Promise.all(promises);`. The main loop waits while those 6 promises resolve (and increment the counter in each then() and then start with the next 6 chunks.

Reply:
Understood, thanks.